flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java

```diff
@@ -307,21 +307,35 @@ private void processBufferOrEvent(BufferOrEvent bufferOrEvent) throws IOExceptio
 			if (event.getClass() != EndOfPartitionEvent.class) {
 				throw new IOException("Unexpected event: " + event);
 			}
+
+			// release the record deserializer immediately,
+			// which is very valuable in case of bounded stream
+			releaseDeserializer(bufferOrEvent.getChannelIndex());
 		}
 	}
 
 	public void cleanup() throws IOException {
-		// clear the buffers first. this part should not ever fail
-		for (RecordDeserializer<?> deserializer : recordDeserializers) {
+		// release the deserializers first. this part should not ever fail
+		for (int channelIndex = 0; channelIndex < recordDeserializers.length; channelIndex++) {
+			releaseDeserializer(channelIndex);
+		}
+
+		// cleanup the barrier handler resources
+		barrierHandler.cleanup();
+	}
+
+	private void releaseDeserializer(int channelIndex) {
+		// recycle buffers and clear the deserializer.
+		RecordDeserializer<?> deserializer = recordDeserializers[channelIndex];
+		if (deserializer != null) {
```

Comment:
Instead of introducing another `@Nullable` field, can not we just always clear deserializer? To me it looks like `releaseDeserializer` (with `deserializer.getCurrentBuffer()` + `deserializer.clear()`) should be already an idempotent method - if yes could you

1. just do not set `recordDeserializers[channelIndex] = null;` 
2. rename `releaseDeserializer` to `clearDeserializer`

+ the same for `StreamTaskNetworkInput` class.

Reply:
The reason why setting `recordDeserializers[channelIndex] = null` is that `clear()` currently does not completely release resources .There is no java doc on `RecordDeserializer#clear()`, we don't know whether the method should ensure the complete release of resources. However, in terms of actual implementation, the following resources in `SpillingAdaptiveSpanningRecordDeserializer` have not been released or incompletely: `SpanningWrapper#buffer`, `NonSpanningWrapper#utfByteBuffer`,  `NonSpanningWrapper#utfCharBuffer`.

Reply:
Ok. In that case according to https://stackoverflow.com/questions/29011893/how-to-annotate-an-array-nonnull :
```
private final RecordDeserializer<DeserializationDelegate<StreamElement>> @Nullable[] recordDeserializers;
```
and please add missing null handling as per warnings.

Reply:
I tried to add the `@Nullable` annotate as you said, but the grammar didn't support it, which reported the compilation error "'`@Nullable`' not applicable to type use". In addition, that question on stackoverflow uses `org.eclipse.jdt.annotation.NonNull` instead of the native annotate of JDK.

Reply:
If we add the state check as another comment said (https://github.com/apache/flink/pull/8471#discussion_r302958200) , I don't think it matter whether it's marked or not here.

Reply:
Ok, let's drop this, since this seems to be supported only for eclipse...

based on https://stackoverflow.com/questions/32327134/where-does-a-nullable-annotation-refer-to-in-case-of-a-varargs-parameter

> When you are using a real Java 8 type annotation, i.e. an annotation with @Target(ElementType.TYPE_USE), the outcome in best explained by using the ordinary array declaration:

And based on:

https://stackoverflow.com/questions/4963300/which-notnull-java-annotation-should-i-use

only Eclipse's `@Nulalble` support ` @Target(ElementType.TYPE_USE) `.